### ISSUE TITLE
Add no-dupe-ternary-expressions

### ISSUE THREAD

**Issue by GrayedFox** [eslint/eslint#13744]
**Please describe what the rule should do:**
The rule will catch and mark as errors any ternary operator that has identical left and right hand expressions. It wall also catch and mark as errors identical ternary conditions which necessarily results in unreachable code.

**What new ECMAScript feature does this rule relate to?**
It does not relate to a new ECMAScript feature, but please indulge me a little and read on. I read the updated contribution guidelines but thought the value of this proposed rule might be high enough despite the fact that it's not related to new JS features.

**What category of rule is this? (place an "X" next to just one item)**

- [x] Warns about a potential error (problem)

**Provide 2-3 code examples that this rule will warn about:**

```js
const getFee = (isMember) => isMember ? 2.00 : 2.00
// "Error: identical ternary expressions"

thing === otherThing ? getFee(isMember) && myBoolean : getFee(isMember) && myBoolean
// "Error: identical ternary expressions"
```

**_Edit: removed examples pertaining to no-unreachable-ternary._**

**Why should this rule be included in ESLint (instead of a plugin)?**
I understand that the nature of ESLint is that it's a pluggable linter and after reading the contribution guidelines, new rules are only generally accepted if they relate to new JavaScript standards/functionality. Considering there are already 200+ rules this principle makes a lot of sense.

Even so, I would argue that duplicate ternary expressions (in the form of statements or conditions) are always and necessarily an error, even if the developer doesn't at first realise what the ternary (or chained ternary) operator is doing.

Avoiding unreachable code is already a great thing for a linter to detect and if we assume that developers want to be informed when a ternary operator does the exact same thing whether or not the condition it evaluates is true or false, then this rule covers both those cases and would be very valuable in any project.

The rule is stand alone (although relates to [no-unreachable](https://eslint.org/docs/rules/no-unreachable) and [no-unneeded-ternary](https://eslint.org/docs/rules/no-unneeded-ternary)) and it's hard to imagine a scenario where developers would not want this rule enabled by default.

For these reasons, I would ask that the ESLint core contributors (hello there!) consider accepting this proposal, which (if accepted and taken as a recommended rule) would be a breaking change and need to be bundled in a major release.

**Are you willing to submit a pull request to implement this rule?**
Absolutely 😸 

**Do you have an idea how this rule could be implemented?**
Yes - the implementation would tokenize all expressions as well as all ternary symbols (so [':', '?']) and create an ordered array. Every element inside the array that is exceeded by a '?' token is a condition and every element that is exceeded by a ':' (or if it is the last element in the array) is an expression.

Some exception/logic for handling complex expression (i.e. function calls or boolean expressions) may be needed.

The resulting array could then easily be searched for identical conditions and identical left and right hand expressions.

**Comment by mdjermanovic** [comment]
In my understanding of the policy, a new rule that "Warns about a potential error (problem)" doesn't have to be related to new ECMAScript features. If that's true, maybe we should update the new rule proposal template to clarify this.

**Comment by GrayedFox** [comment]
Just for context, this is a [reboot of an old proposal](https://github.com/eslint/eslint/pull/12102#partial-pull-merging) that ended up getting way out of scope and having too many overlaps with existing rules.

This time round I've got a much clearer idea about the implementation and hope this is a well rounded, more precise proposal that keeps to itself 😉

**Comment by mdjermanovic** [comment]
> ```js
> const getFee = (isMember) => isMember ? 2.00 : 2.00
> // "Error: identical ternary expressions"
> 
> return condition1 ? value1 : condition1 ? value2 : value3
> // "Error: identical ternary conditions (unreachable code)"
> ```

I think these should be two rules since they're checking two different things: 

* One for identical "branches".
* The other for conditions, like `no-dupe-else-if`.

> (if accepted and taken as a recommended rule) would be a breaking change and need to be bundled in a major release.

We can add a new rule in a minor version, and make it recommended in a major version.

**Comment by anikethsaha** [comment]
The proposal looks good to me, Agreed that these can be two different rules 

About the unreachable code

like here

```js
return condition1 ? value1 : condition1 ? value2 : value3;
```

if we have something like

```js
return i++ < 10 ? value1 : i++ < 10 ? value2 : value3;
```

Not sure whether this would be noted as unreachable or not. But if this is noted as unreachable then I guess it would be a false case as it won't be solved using static analysis. We can document this as a known limitation.

**Comment by GrayedFox** [comment]
Cool. So, seperate rules and as for names how about `no-dupe-ternary-expressions` and `no-dupe-ternary-conditions`? 

> ...if we have something like
> 
> ```js
> return i++ < 10 ? value1 : i++ < 10 ? value2 : value3;
> ```

Hmm, interesting. I think that value2 in this case is still unreachable code (even with a prefix increment i.e. ++i). I originally thought there would be [quite a few outliers](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators) that might result in a false positive but I just tried a combination of different bitwise shift operators, assignment operators, and increment/decrement operators in the Firefox console and only the increment/decrement operators didn't throw a syntax error.

I could be wrong (I just did `fn = (i) => (++i < 10) ? 'value1' : (++i < 10) ? 'value2' : 'value3'` and logged the result of calling the function with different values, as well as reassigned the function using `i**`, so not exactly exhaustive testing!) but I cannot think of any identical conditions now that don't result in unreachable code - based on Firefox (or the V8 engine, I guess) rejecting bitwise and exponent operators as part of the ternary condition.

Keen to get some more eyes on this just in case!

**Comment by mdjermanovic** [comment]
It's possible to write some code with side effects where this would be a false positive, but it's so rare that we can add a "When Not To Use"  section for such cases, like in [no-dupe-else-if](https://eslint.org/docs/rules/no-dupe-else-if#when-not-to-use-it).

**Comment by mdjermanovic** [comment]
> I actually have doubt whether this will be unreachable by eslint (code path analysis)?

No, code path analysis doesn't compare expressions or check values except for a couple of cases where it isn't so unusual to have constants, like `while (true)`.

**Comment by anikethsaha** [comment]
Cool thanks,

+1 for separated rules with noted limitations as discussed above.

**Comment by GrayedFox** [comment]
Great, than I will get to work on this tomorrow, thanks for the fast feedback

**Comment by mdjermanovic** [comment]
Please note that new rules need [consensus](https://eslint.org/docs/maintainer-guide/issues#accepting-issues) from the team to be accepted.

Let's wait for some more opinions, and if it turns out that these should be two different rules then we should have two issues to evaluate them separately.

**Comment by GrayedFox** [comment]
I did get started but I understand that until the rule is championed and consensus is reached that there is no guarantee it will be accepted or merged 🙏🏾 (just wanted to take advantage of the weekend!)

**Comment by nzakas** [comment]
Our official (and documented) policy is that new rules must be related to new language features. Unofficially we *might* consider new rules not related to new language features if the errors they unveil are serious enough.

To me, this proposal doesn’t meet the bar of “serious enough.” The downside of not flagging this pattern is that everything works. This is flagging an efficiency issue, not really an error.

This seems exactly the case where a plugin makes sense.

**Comment by GrayedFox** [comment]
To be honest, I think the "everything works" statement is worth unpacking a bit :wink:

IMHO debugging this sort of problem - something that a linter is more than capable of detecting without more than static code analysis - is precisely the kind of thing that warrants addition to the core rule set. Everything works (in the best case scenario) means something like "the code behaves as expected, but the ternary expression is unnecessary" (which is also confusing and especially so if you're the one reading someone else's code).

Worst case scenario is everything works but since no linting error or warning is thrown, depending on how the project and codebase is set up, what could simply have been an accidental keystroke costs the team time and resources tracking down. Speaking from experience, I was surprised after debugging some code that our linter (in this case a ruby project) hadn't detected the identical ternary conditions (where this idea came from).

I agree with you that this is certainly something that could be plugged in instead of added to the core rules. Perhaps, given that for some this could be more of an efficiency issue, the default rule could be to warn instead of an error? Does that fit in with the ESLint standards/norms?

**Comment by GrayedFox** [comment]
That said, I did say that I understood that without consensus there was no guarantee of a merge or implementation, and I did also read the docs about new rules relating to new language features, so I will of course accept that outcome if that is the case 🙇🏾‍♂️

**Comment by nzakas** [comment]
@GrayedFox I don't want to get too far into the weeds here, but the bottom line is that the bar a rule has to meet to be included in the core is necessarily high because we just don't have the bandwidth to keep accepting new rules just because they have some utility to some people. If this same rule was proposed four years ago, we'd probably accept it without much concern. As we approach 300 rules, with people expecting no overlapping warnings, we just can't do that anymore. There are a bunch of core rules that wouldn't be accepted today if they were proposed as new...that's just the way things are.

Again, this is precisely why we focused so much on creating a robust plugin system and ways to share rules. I never wanted the ESLint team to be the ultimate gatekeeper for what rules make sense to exist and use.

**Comment by GrayedFox** [comment]
It's all good :) there's no bad blood from my end. I published the proposed rules (and already closed the PRs) as a plugin, I will also close this issues now, I appreciate that it was considered.

Ultimately I still got to make use of the work by using the plugin system as you recommended, so I'm happy, was a great 20 hours or so of work during lockdown times (now it's back to C# land for a bit).

For those interested: https://www.npmjs.com/package/eslint-plugin-ternary
